actor/builtin/miner/miner.go

```diff
@@ -714,3 +723,75 @@ func (ma *Actor) GetProvingPeriodStart(ctx exec.VMContext) (*types.BlockHeight,
 
 	return state.ProvingPeriodStart, 0, nil
 }
+
+func (ma *Actor) currentProvingPeriodPoStChallengeSeed(ctx exec.VMContext) (proofs.PoStChallengeSeed, error) {
+	chunk, err := ctx.ReadStorage()
+	if err != nil {
+		return proofs.PoStChallengeSeed{}, err
+	}
+
+	var state State
+	if err := actor.UnmarshalStorage(chunk, &state); err != nil {
+		return proofs.PoStChallengeSeed{}, err
+	}
+
+	bytes, err := ctx.SampleChainRandomness(state.ProvingPeriodStart)
+	if err != nil {
+		return proofs.PoStChallengeSeed{}, err
+	}
+
+	seed := proofs.PoStChallengeSeed{}
+	copy(seed[:], bytes)
+
+	return seed, nil
+}
+
+// SampleChainRandomness produces a slice of random bytes sampled from a TipSet
+// in the blockchain at a given height (minus lookback). This is useful for
+// things like PoSt challenge seed generation.
+//
+// If no TipSet exists with the provided height (sampleHeight), we instead
+// sample the genesis block.
+func SampleChainRandomness(sampleHeight *types.BlockHeight, tipSetCh <-chan interface{}) ([]byte, error) {
+	// EDGE CASE: for now if ancestors includes the genesis block we take
+	// randomness from the genesis block.
+	// TODO: security, spec, bootstrap implications.
+	// See issue https://github.com/filecoin-project/go-filecoin/issues/1872
+	sampleHeight = sampleHeight.Sub(types.NewBlockHeight(LookbackParameter))
+	if sampleHeight.LessThan(types.NewBlockHeight(0)) {
+		sampleHeight = types.NewBlockHeight(0)
+	}
+
+	var sampleTipSet *types.TipSet
+
+Loop:
+	for raw := range tipSetCh {
+		switch v := raw.(type) {
+		case error:
+			return nil, xerrors.Wrap(v, "error walking chain")
+		case types.TipSet:
+			height, err := v.Height()
+			if err != nil {
+				return nil, xerrors.Wrap(err, "error obtaining tip set height")
+			}
+
+			if sampleHeight.Equal(types.NewBlockHeight(height)) {
+				sampleTipSet = &v
+				break Loop
```

Comment:
If I use `continue` instead of `break Loop`, this code will always traverse the blockchain back to the genesis block. We don't want to do that if we can avoid it.

Reply:
Sorry, I misunderstood break label.

Reply:
I forgive you my son